src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java

```diff
@@ -352,6 +286,24 @@ public static FunctionItem getUserDefinedFunction(FunctionIdentifier identifier,
          * function that parallelizes item collections into a Spark RDD
          */
         public static final String PARALLELIZE = "parallelize";
+        public static final BuiltinFunction parallelizeFunction1 = new BuiltinFunction(
```

Comment:
I have a productivity comment, to help with the easy population of the builtin function catalogue.

Here, it may be useful to factor out the declarations inside utility methods, inside convenience functions. Then we could have a very compact list list of declarations, looking like this for the four methods already appearing here:

putBuiltInFunction(PARALLELIZE, ParallelizeFunctionIterator.class, "item*", "item*");
putBuiltInFunction(PARALLELIZE, ParallelizeFunctionIterator.class, "item*", "integer", "item*");
putBuiltInFunction(ENDSWITH, EndsWithFunctionIterator.class, "string?", "string?", "boolean");
putBuiltInFunction(HEXBINARY, HexBinaryFunctionIterator.class, "string?", "hexBinary?");

with overloaded putBuiltInFunction methods (one for each arity between 1 and 4, I don't think we have builtin functions for more than 4 -- and the arity is implicitly passed via the number of arguments). The methods putBuiltInFunction can then take care of everything: creating the BuiltinFunction objects as well as calling builtInFunctions.put().

That way, it would be very easy and quick to create the entire list of signatures: these would be all one liners.

Does it help?

Reply:
Dear Prof. Fourny, 
I refactored the whole builtin functions catalogue with the necessary types. Please let me know if the way I implemented it looks good or you prefer to have the put-into-map inside the same function that creates the builtin function. I kept the two things separate but I have no strong preference for one over the other. 

There are quite some file changes because, as I went over the different iterators, I did some code refactoring when necessary to have a cleaner code. 
I also fixed the substring iterator that was not behaving according to specs